### PR TITLE
Implement boss bars

### PR DESCRIPTION
**Pull request by Alfalik**
I've coded a basic boss bar for time remaining, here's a video of how it works and looks (https://youtu.be/X7fEoY2Mol0). I'll be thankful if someone help contributing to any of these inconsistencies:

+ Wither being visible sometimes
+ Bar flashing occasionally

If you find any problem(s), don't forget to reply here, thanks.

### CODE REVIEW

**Comment by dentmaged** [comment]
Awesome @Alfalik!

**Comment by Prodigy4532** [comment]
Nice!!

**Comment by iEli2tyree011** [comment]
This looks okay, but it seems that you made it so the code is very heavily based around the actual Time Remaining boss bar. I feel a system should be managed to keep track of all boss bars, but only display the one that will end the soonest. Then, multiple boss bars can be integrated into that. @twizmwazin may have other ideas, but as I said earlier, it looks pretty good, but almost seems like the Time Remaining bar is the priority, while there are several other messages that will be displayed on the boss bar that we have to keep in mind.

**Comment by Pablete1234** [comment]
In OCN you can also see the wither in the distance, if the wither's health is low

**Comment by Alfalik** [comment]
@iEli2tyree011 I just believe that the Time Remaining bar is the most important and the unique that should be in Cardinal. I don't think that start or cycle bars are useful, even though that they aren't that difficult to code. I'll leave the decision to @twizmwazin.

@Pablete1234 Thanks for the info! I'll see if it can be fixed

File: src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java [hunk 1]
```diff
@@ -75,45 +88,127 @@ FakeWither getWither(Player player, String message) {
...
+        if (visible) {
+            ((CraftPlayer) player).getHandle().playerConnection.sendPacket(wither.getSpawnPacket());
+        }
         players.put(player, wither);
         return wither;
     }
 
     private Location getWitherLocation(Player player) {
-        return player.getLocation().add(player.getEyeLocation().getDirection().multiply(100));
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
```
Comment (ryanw-se) [inline src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java]: @Alfalik What is the point of this if/else statement? You're returning the same thing for both cases, if this is the case you don't even need the statement.

File: src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java [hunk 2]
```diff
@@ -75,45 +88,127 @@ FakeWither getWither(Player player, String message) {
...
+        }
+        else {
+            return player.getLocation().add(player.getEyeLocation().getDirection().multiply(110));
+        }
+    }
+
+    private void destroy(Player player) {
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            FakeWither wither = players.get(player);
+            if (wither == null)
```
Comment (ryanw-se) [inline src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java]: @Alfalik Your return statement can probably be consolidated to one line like so, ```if (wither == null) return;```java

File: src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java (hunk 1 above)
Comment (Alfalik) [inline src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java]: @ryanw-se Oh, it was supposed to be different, I'll correct it, thanks.

File: src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java [hunk 3]
```diff
@@ -75,45 +88,127 @@ FakeWither getWither(Player player, String message) {
...
     @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     public void onPlayerRespawn(final PlayerRespawnEvent event) {
-        handleTeleport(event.getPlayer(), event.getRespawnLocation().clone());
+        if (GameHandler.getGameHandler().getMatch().isRunning()) {
+            if (players.get(event.getPlayer()) != null) {
+                handleTeleport(event.getPlayer(), event.getRespawnLocation().clone(), true);
+            }
+
+        }
+        if (TimeLimit.getMatchTimeLimit() == 0) {
```
Comment (ryanw-se) [inline src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java]: @Alfalik Your if statement checking if the match time limit is zero and then deleting the boss bar is a bit repetitious, you should consolidate it down into one statement in the handleTeleport method.

File: src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/FakeWither.java [hunk 4]
```diff
@@ -40,10 +41,6 @@ public float getMaxHealth() {
         return maxHealth;
     }
 
-    public void setMaxHealth(float max) {
-        maxHealth = max;
-    }
-
```
Comment (ryanw-se) [inline src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/FakeWither.java]: @Alfalik What was the point of moving this method? Its purely an aesthetic decision at this point and shouldn't effect your PR in any way. Same for the movement of the variables above, which were structured that way for a reason. Public variables were on top and the private ones were on the bottom, there is no point in changing this.

**Comment by ryanw-se** [comment]
I did a quick review, its decent. Personally I feel the implementation is built around the time remaining bar a tad too much and would make it more challenging to implement other bars in the future. A good implementation should have support and make it easier for future boss bars to be added.

**Comment by Alfalik** [comment]
@ryanw-se Thanks for the review and taking the time for do it. Now, I'll try to get the time remaining, start and cycle bars, all working. Thanks again

**Comment by twizmwazin** [comment]
I'd like to see a bar system, where a Bar object can be created, and players can have a Set<Bar>. The bossbar module would then use those bars, displaying the one with the smallest time remaining.

**Comment by Pablete1234** [comment]
@twizmwazin What do you mean by "how overcast does their wither so far away", they have it about 50-60 blocks away. and yes, towards the end, you can see it in the distance

**Comment by TiTo-418** [comment]
https://oc.tc/forums/topics/55a824e567d52260e60041f9

**Comment by Pablete1234** [comment]
Full info about the wither: http://pastebin.com/1rBhVKBt , and it's located 50 blocks away from the player

**Comment by TiTo-418** [comment]
@Pablete1234 I think the withers can't have invisibility

**Comment by Pablete1234** [comment]
@TiTo-418 But it's not using an "Age" tag either :/

**Comment by twizmwazin** [comment]
@Pablete1234 That isn't going to work. Besides the fact that vanilla commands are entirely useless in Cardinal, the wither needs to be far from the player. The implementation system doesn't actually create a wither, it only sends packets to simulate a wither to the client. Because the Minecraft server doesn't know there is a wither we don't have to mess with any vanilla functionality.

**Comment by Pablete1234** [comment]
@twizmwazin I know it's only sending the packets to do so, but if the client recives a wither, with invul 880, NoAI 1, and 50 blocks away, the client will not see it bocs it's so small, and particles don't render at that distance

**Comment by twizmwazin** [comment]
@Pablete1234 No, it won't. The command tells the server how to handle the wither. In our case, there is no such wither.

**Comment by Pablete1234** [comment]
@twizmwazin Even tho the wither never exists for the server, it gives an idea of what the packets sent to the player should be, now go ahead and create a packet that sends the same wither as a vanilla server having one does.

**Comment by twizmwazin** [comment]
@Pablete1234 None of that information is even slightly useful when creating the packets. Creating packets is really straightforward. I've already implemented a bossbar before, although it didn't meet my standard, so I scrapped it. The issue is finding how Overcast has the wither render on the client despite being out of the render distance.

**Comment by Alfalik** [comment]
@twizmwazin The max rendering distance I've tried is 112 approx., if the distance is 113, the wither won't render. Can't understand how OCN sends his wither ~200 blocks far. I'd be happy if someone find the way to.

**Comment by Alfalik** [comment]
These exact particles need to be removed when ```percent < 50%```. I don't know if remove particles from a mob is a thing.

**Comment by ryanw-se** [comment]
@Alfalik You can't remove the shine effect the mob gets when it reaches low health, its not possible to get rid of it unfortunately, if you go into a game on the Overcast Network you can see it in your cross hairs as well, (they didn't get rid of it either).

**Comment by ryanw-se** [comment]
@Alfalik As for the max wither distance, I'm assuming it has something to do with server render distance and the client one, at 112 you're server render distance and possibly your client one would have to be set to ``7```.

**Comment by Pablete1234** [comment]
@twizmwazin Again: overcast does not have the wither so far away, they have it 50 blocks away from the player, the Invul tag makes it smaller, find the way to include that in a packet

**Comment by Pablete1234** [comment]
Anyone that has a tm server, do the test: 
 - Get in gamemode 1, and fly (so you don't fall)
 - /tl set 1h (to get a bar)
 - /minecraft:tp ~ ~45 ~
 - Do not move the camera, and you will see the wither's particles

**Comment by CaptainElliott** [comment]
@Pablete1234 twiz has tried to explain this to you, that's not how packets works, so stop saying "find a way to include that in a packet"

**Comment by Pablete1234** [comment]
@CaptainElliott But everyone is trying to find out how far they can set the wither, it's not how it works on ocn!

**Comment by Alfalik** [comment]
Okay. I've made it more "globally". The thing that I need for the formula for start/cycle boss bar and can't find the way to, is the joinedTime. For example, a player do ```/start 25```, the ```joinedTime``` will be 25, and the ```time``` 25. After 10 seconds, the ```joinedTime``` will be 25 and the ```time``` will be 15. With the joinedTime I'll do the formula, that seems like this: ```float percent = (int) (time * 100F) / joinedTime```. That's what I need for start and cycle boss bars.

**Comment by CaptainElliott** [comment]
@Pablete1234 how can you be so sure when you have no experience  with bukkit and it's packets?

**Comment by Pablete1234** [comment]
@CapitainElliott I have 0 experience with bukkit, but go to ocn and do the tp thing, and you will see i'm not wrong about the 50 blocks thing, and I know for sure, that the client is rendering a wither with Invul tag

**Comment by dentmaged** [comment]
@Pablete1234 There is no wither for the tag to be applied. It's an "illusion". Only the player can see the wither because it's sent using particles.

**Comment by Pablete1234** [comment]
@dentmaged well, the "illusion" needs to include the tag, ocn does it, so it's not impossibe

**Comment by dentmaged** [comment]
@Pablete1234 And how do you know OCN does it?

**Comment by Pablete1234** [comment]
@dentmaged because the wither is upsidedown, that's what high Invul does

**Comment by Alfalik** [comment]
Here's how OCN does his own boss bar (https://youtu.be/sxoA4ZIgba0).

**Comment by Pablete1234** [comment]
@Alfalik that's exactly what i said... and everyone is saying "you can't do that with packets"

**Comment by twizmwazin** [comment]
Ok, I see what it is now. Although I still don't know what "invul" means, which will be necessary to make the packet. There also needs to be a way to have it show regardless of render distance.

@ryanw-se The server's "render" distance is a server-side setting to tell the server what packets should be sent to the client. Once received by the client, it is up to the client to handle the packets and render the world.

**Comment by Pablete1234** [comment]
@twizmwazin If you spawn a wither in survival, the time it takes to explode, is saved in the Invul tag (it's ticking down), when you just spawn the wither, it's small, and becomes bigger as the invul ticks down, when it reaches 0, the wither explodes and "the battle starts" (if you are in survival), when setting this value higher than it usually is, it makes the wither smaller and smaller, to the point where it's invisible (880) and even upside down (what ocn uses: 1000)

**Comment by Alfalik** [comment]
@twizmwazin This is what invul does (https://youtu.be/h1mqy1BaitU), but sportbukkit don't have anything related to invul. Here's the command if needed ```/summon WitherBoss x y z {NoAI:1,Silent:1,Invul:1000,Health:80F}```.

**Comment by Pablete1234** [comment]
@Alfalik noo!! what did you do!! vanilla?? never use vanilla!?!?!?!? you may say "But now I know how invul works and affects the wither" well:
"No, it won't. The command tells the server how to handle the wither. In our case, there is no such wither. Also, you need to take into account render distance."

**Comment by twizmwazin** [comment]
@Pablete1234 Really? Be mature. The vanilla command doesn't help me much at all, however an understanding of this "invul" tag does. Unless you are the one actually writing this and know what you are doing, I'd advise minding your own buisness.

**Comment by Pablete1234** [comment]
@twizmwazin maybe i'm being immature, but I've been talking about that for 2 days and everyone has ignored me instead of trying to run my tests, @Alfalik just recorded what i had said, and now you understeand, but you didn't care at all of what I said

**Comment by twizmwazin** [comment]
Even if the it is the same information, Alfalik presented it in a way I could understand. I really couldn't figure out what you were trying to say other than that you could use a vanilla command to summon a wither.

**Comment by Aaron1011** [comment]
The `Invul` tag data is sent to the client through entity metadata: http://wiki.vg/Entities#Wither

**Comment by Alfalik** [comment]
@Aaron1011 Thanks! That's what I needed

**Comment by twizmwazin** [comment]
An implementation written by someone else that seems to work fairly well from my brief testing. https://bitbucket.org/inventivetalent/bossbarapi/src/407c3fc2e5c887065c35596d9e382c895ed3d188/BossBarAPI/src/org/inventivetalent/bossbar/BossBar.java?at=master

**Comment by twizmwazin** [comment]
I have a handful of things I want to do for bossbars, I'm going to try and get to them, this may need to work around them.

**Comment by Alfalik** [comment]
@twizmwazin I've got to work start, cycle and time remaining boss bars. But, a NPE is thrown when they are sent. Also, need to find the right formula to get percent on start and cycle boss bars. Hopefully you can test the changes.

**Comment by Pablete1234** [comment]
@Alfalik when someone starts a countdown, save the total time, and then percentatge = time remaining / total time * 100

**Comment by Alfalik** [comment]
@Pablete1234 I know the formula, the unique problem is how to get the joined time, or as you said, total time.

**Comment by dentmaged** [comment]
@Alfalik https://github.com/Alfalik/CardinalPGM/blob/2c3b64127bfbe8a4dbf49cbdeeec1725a5c48b40/src/main/java/in/twizmwaz/cardinal/module/modules/cycleTimer/CycleTimerModule.java#L23

Just create another integer named originalTime then add ```setOriginalTime(time);``` in this method: https://github.com/Alfalik/CardinalPGM/blob/2c3b64127bfbe8a4dbf49cbdeeec1725a5c48b40/src/main/java/in/twizmwaz/cardinal/module/modules/cycleTimer/CycleTimerModule.java#L102

**Comment by Alfalik** [comment]
@dentmaged Thanks, it works.

The unique and actual problem is a NullPointerException thrown when a boss bar is created. I don't know what's wrong with the code or wheres the error. Hope anyone can help me.

**Comment by dentmaged** [comment]
@Alfalik You're welcome. I encountered the same issue whilst working on Cardinal+. I can't remember how I fixed it. Just use ```System.out.println(something == null);```

**Comment by Alfalik** [comment]
A quick update:

+ The NPE were fixed by @dentmaged.
+ Add a description after every metadata of the wither.
+ Remove isState and use getState() == State instead.
+ Int originalTime is everytime 30, I'll fix it as soon as possible.

**Comment by Pablete1234** [comment]
@Alfalik the wither needs to go closer to the player, playing at 5 chunks it flashes, and less than 4 chunks it won't render at all.

**Comment by twizmwazin** [comment]
Did some brief testing, I'm pretty impressed with this so far. Besides Pablete's issue, the only other issue I noticed was that you assume that the start countdown is 30. You will need to make the max whatever the user entered as the start time.

**Comment by Alfalik** [comment]
As @Pablete1234, the wither distance were changed to 50 blocks from the players. About the match start message, it is a half of a second, I don't see other way to do it or the problem itself with it.
@twizmwazin Still trying to fix ```originalTime``` on start timer being 30.

**Comment by twizmwazin** [comment]
You probably want to put the wither around 20 blocks away, so it can be seen from the smallest render distance.

**Comment by Alfalik** [comment]
@twizmwazin Done. I'll be working on the start issue tomorrow.

**Comment by Pablete1234** [comment]
@Alfalik the 1 second remaining message should stay for 1 second, and the match starting now should be the "0 seconds remaining" message (if there was one) and should stay until 1 second of the match has passed.
ps: In OCN they removed the "match starting now", it's just "match starting in 0 seconds" , and it doesn't even show up if it's a match with time limit, it's like lowest priority

**Comment by Alfalik** [comment]
@Pablete1234 I just removed it. Thanks for meticulous testing although.

**Comment by Alfalik** [comment]
Yess

**Comment by ryanw-se** [comment]
@Alfalik Nicely done! :clap:

**Comment by Pablete1234** [comment]
@Alfalik :clap:

**Comment by Alfalik** [comment]
@Pablete1234 The ```show-before="time"``` is only for monument modes, and it will show the one with less time.
About the click sounds, could you find the exact sound, volume and pitch of it? That'll be helpful.

**Comment by TiTo-418** [comment]
@Alfalik Look at https://github.com/twizmwazin/CardinalPGM/issues/227

**Comment by Alfalik** [comment]
Update:

+ Got to work the monuments modes bar, but it still have some errors:
    - It doesn't work without time limit.
    - Need to reset time of timeBeforeMode after every monument mode runs. That's why (if anyone test) only the first mode is shown in the boss bar.
+ On start timer, ```originalTime``` always return 30 seconds.

If somebody could help with any of these inconsistencies, it would help much.

@TiTo-418 It isn't there.

**Comment by iEli2tyree011** [comment]
@Alfalik Monument modes always have a name

**Comment by dentmaged** [comment]
@Alfalik When monuments have no name define the XML it is the blocks in uppercase without underscores in the colour red.

**Comment by Alfalik** [comment]
@iEli2tyree011 I'm not that sure, look at docs https://docs.oc.tc/modules/monument_modes
@dentmaged Thanks

**Comment by iEli2tyree011** [comment]
@Alfalik 
> Monument modes always have a name

**Comment by Alfalik** [comment]
Well, the monument modes boss bar totally works, except if the match does not have a time limit. I still can't understand the start timer issue that return always 30 seconds. The rest is working perfectly fine.

**Comment by AngelCasillas01** [comment]
Maybe this works in cannon duel too

**Comment by TiTo-418** [comment]
@AngelCasillas01 
> Well, the monument modes boss bar totally works, *except if the match does not have a time limit*.

**Comment by AngelCasillas01** [comment]
@TiTo-418 Comfirmando :p

**Comment by Prodigy4532** [comment]
@AngelCasillas01 "Confirmando"*

**Comment by Alfalik** [comment]
A quick update. I've been making the boss bar more consistent (solving bar health flashing and the wither being visible sometimes), things that have been totally fixed. I still have two issues that really not have reviewed:

+ Int ```originalTime``` on Start Timer always return to 30 seconds.
+ Monument modes bar not working if the match doesn't have a time limit.

Hope anyone can help or tell me some issues (if found).

**Comment by TiTo-418** [comment]
@Alfalik Can you, with the boss bar active, do /cycle 0? I think you will find an issue, i can not describe it

**Comment by Alfalik** [comment]
@TiTo-418 Could not find the issue. Can you explain it?

**Comment by TiTo-418** [comment]
@Alfalik I saw the wither particles, it is very fast. Im not sure, maybe im crazy. I'll try to record it.

**Comment by Alfalik** [comment]
@TiTo-418 Yes. Those particles are sent when the wither die.

**Comment by TiTo-418** [comment]
@Alfalik The boss bar disappear if you die and respawn

**Comment by Alfalik** [comment]
@TiTo-418 Could you ```git clone``` again and test, I've ammend the commit

**Comment by TiTo-418** [comment]
@Alfalik Sure

**Comment by TiTo-418** [comment]
@Alfalik Still happen

**Comment by twizmwazin** [comment]
I'm taking a look at this, couple of things:
 - Use proper package naming conventions. It is standard to use camelCase, you have a package inproperly named `Reflection`. Regardless this would be more suited in util.
 - Wildcard imports. I don't like them usually, and importing the whole Minecraft server is a prime example of why not. It may have been an accidental thing done automatically by your IDE, but it should be configurable, I know for sure it is in IntelliJ.
 - Back on reflection, I don't really want to use reflection when it isn't necessary. It is ok when accessing otherwise private fields and methods, although we don't need it for NMS classes. We can just access them directly, it is already done by the plugin.

**Comment by Alfalik** [comment]
Thanks for taking a look, and about the things that you pointed:

+ I never change/touch imports, IntelliJ replace them itself with wildcards, automatically, as you said. I've disabled this function altough.
+ Should I move reflection to util and leave it for future things, or directly remove it? I find it useful and easy to understand.

**Comment by twizmwazin** [comment]
I'd just remove the reflection classes altogether.

**Comment by Alfalik** [comment]
@twizmwazin Done
@TiTo-418 The wither disappearing issue if a player dies have been totally fixed.

**Comment by Alfalik** [comment]
I've finished implementing all boss bars (start, cycle, time remaining and monument modes). The two issues that I had, are totally fixed. I'd be happy if anyone can further test and tell me if it is completely ready to be merged or need some enhancements.

**Comment by TiTo-418** [comment]
@Alfalik The particles problem is not fixed

**Comment by TiTo-418** [comment]
<a href="http://imgur.com/r3xPj7Q"></a>

**Comment by TiTo-418** [comment]
Sometimes it flashes

**Comment by Alfalik** [comment]
@TiTo-418 About your points:

+ It isn't a problem.
+ I'd recommend to test with a server, such as hosthorde, with a minium of 2 GB of RAM, not a localhost. Also define how it flashes for you, the bar return to the lower health or it dissapear totally.
+ It appears for me.
+ And for the image, I'll take a look, I misspelled something.

**Comment by Alfalik** [comment]
@TiTo-418 Could you test again?

**Comment by TiTo-418** [comment]
@Alfalik It works perfect now :D

**Comment by Alfalik** [comment]
@twizmwazin Any thoughts?

**Comment by twizmwazin** [comment]
I can do some testing in a bit, do you think this is ready to be merged?

**Comment by TiTo-418** [comment]
@twizmwazin Yes, it was tested today and works fine

**Comment by Alfalik** [comment]
@twizmwazin Yes, I've fixed all the errors or inconsistencies.

**Comment by Pablete1234** [comment]
@Alfalik OCN uses one wither and keeps changing the name, health and position of that one, instead of summoning a new one each time

**Comment by Alfalik** [comment]
@Pablete1234 That's how this boss bar works. The wither is only sent on this events https://github.com/Alfalik/CardinalPGM/blob/bossbar/src/main/java/in/twizmwaz/cardinal/module/modules/bossBar/BossBar.java#L144, and the wither is updating constantly.

**Comment by Pablete1234** [comment]
If you /ready and then /unready the wither stays there

**Comment by Alfalik** [comment]
Try now

**Comment by twizmwazin** [comment]
I'm under the impression that this is ready to be merged, @Alfalik can you rebase this?

**Comment by Alfalik** [comment]
@Pablete1234 The monuments boss bar staying until the match end issue, is now fixed.
@twizmwazin Done!

**Comment by Pablete1234** [comment]
The commit message deserves to be "Implement a boss bar system" it's not basic anymore

**Comment by Pablete1234** [comment]
I want to point out that ocn does not have /tl set 1h, ocn only has /tl 1h, and means in the moment you type the command, the countdown starts going, that means after a /tl command the bossbar always gets to full health
In cardinal, you can /tl set 1h, when 50 mins have passed, and the bossbar won't start at  max health, ti will start at 10/60 mins

I don't know what's the prefered behaviour, but i'll let you know guys

**Comment by dentmaged** [comment]
@Pablete1234 When @twizmwazin (or @iEli2tyree011) was writing the ```/timelimit``` command, he asked if it should be ```/timelimit <time>``` or ```/timelimit <set|add>```. I prefer ```/timelimit <time>```, I might rewrite it in a PR (if I get enough time).

**Comment by Pablete1234** [comment]
@dentmaged but with /tl set you can choose exactly how long the match is going to be, for example, if a match starts with no tl, and at the mark of 50 mins you realize you want to set the tl, in ocn you can't ensure the match will end at the 1h mark exactly, in cardinal you can
Maybe the best behaviour would be that /tl set 60m on 51:24 is an "alias" of /tl 8m36s, and /tl add 10m (just after the first one) is an "alias" of /tl 18m36s

**Comment by dentmaged** [comment]
@Pablete1234 The ```/timelimit``` command is buggy and needs to be rewritten. The time remaining messages are misfiring after setting the time limit.

**Comment by Pablete1234** [comment]
@dentmaged i know, but that's more about the time notifications, they fire when you /tl, and they schedule themselves to fire again 5 mins later, instead of firing on the command, and on the next 5 multiple time

**Comment by twizmwazin** [comment]
merged

**Pull request closed without merging**